File: src/main/scala/com/vc/medium/L648.scala
package com.vc.medium

object L648 {
  def replaceWords(dict: List[String], sentence: String): String = {
    import scala.collection.mutable
    case class Node(child: mutable.HashMap[Char, Node], var isLeaf: Boolean)
    val root = Node(new mutable.HashMap[Char, Node], isLeaf = false)
    var i = 0
    while(i < dict.length) {
      val word = dict(i)
      var j = 0
      var current = root
      while(j < word.length) {
        val ch = word(j)
        if(!current.child.contains(ch)) {
          val node = Node(new mutable.HashMap[Char, Node], isLeaf = false)
          current.child.put(ch, node)
        }
        if(j == word.length - 1) current.child(ch).isLeaf = true
        current = current.child(ch)
        j += 1
      }
      i += 1
    }

    val str = new mutable.StringBuilder()
    i = 0
    val words = sentence.split(" ")
    while(i < words.length) {
      val newWord = new mutable.StringBuilder
      val word = words(i)
      var current = root
      var flag = false
      var j = 0
      while(j < word.length) {
        val ch = word(j)
        if(current.child.contains(ch)) {
          newWord.append(ch)
          if(current.child(ch).isLeaf) {
            flag = true
            str.append(newWord).append(" ")
            j = word.length
          }
          current = current.child(ch)
        }
        else j = word.length
        j += 1
      }
      if(!flag) str.append(word).append(" ")
      i += 1
    }
    str.substring(0, str.length - 1).toString
  }
}

File: src/main/java/com/vc/medium/L648.java
package com.vc.medium;

import java.util.*;

class L648 {
    static class Node {
        HashMap<Character, Node> map;
        boolean isWord;
        Node() {
            map = new HashMap<Character, Node>();
        }
    }

    class Trie {
        Node root = new Node();

        private void addWord(String word) {
            Node current = root;
            for(int i = 0; i < word.length(); i++) {
                char ch = word.charAt(i);
                if(!current.map.containsKey(ch)) current.map.put(ch, new Node());
                current = current.map.get(ch);
            }
            current.isWord = true;
        }

        private String getParent(String word) {
            Node current = root;
            String w = "";
            for(int i = 0; i < word.length(); i++) {
                char ch = word.charAt(i);
                if(current.isWord) return w;
                if(!current.map.containsKey(ch)) return null;
                current = current.map.get(ch);
                w += ch;
            }
            return null;
        }
    }

    public String replaceWords(List<String> dict, String sentence) {
        Trie trie = new Trie();
        for(String d: dict) trie.addWord(d);

        StringBuilder sb = new StringBuilder();
        for(String s: sentence.split(" ")) {
            String parent = trie.getParent(s);
            if(parent == null) sb.append(s);
            else sb.append(parent);
            sb.append(" ");
        }
        return sb.deleteCharAt(sb.length() - 1).toString();
    }

}
